pkg/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,343 @@
+	// NewNetworkPolicyController.
+	anpLister seclisters.NetworkPolicyLister
+	// anpListerSynced is a function which returns true if the AntreaNetworkPolicies shared informer has been synced at least once.
+	anpListerSynced cache.InformerSynced
+}
+
+func NewStatusController(antreaClient antreaclientset.Interface, internalNetworkPolicyStore storage.Interface, cnpInformer secinformers.ClusterNetworkPolicyInformer, anpInformer secinformers.NetworkPolicyInformer) *StatusController {
+	c := &StatusController{
+		npControlInterface: &networkPolicyControl{
+			antreaClient: antreaClient,
+			anpLister:    anpInformer.Lister(),
+			cnpLister:    cnpInformer.Lister(),
+		},
+		queue:                      workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "networkpolicy"),
+		internalNetworkPolicyStore: internalNetworkPolicyStore,
+		statuses:                   map[string]map[string]*controlplane.NetworkPolicyNodeStatus{},
+		cnpListerSynced:            cnpInformer.Informer().HasSynced,
+		anpListerSynced:            anpInformer.Informer().HasSynced,
+	}
+	// To save a "GET" query before each update, UpdateAntreaClusterNetworkPolicyStatus treats the cache of Lister as
+	// the state of kube-apiserver. However, in some case the cache may not be in sync, then we might skip updating a
+	// policy's status by mistake. To resolve it, add update event handlers which triggers resync of a policy if it's
+	// status is updated.
+	cnpInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.updateCNP,
+		},
+		resyncPeriod,
+	)
+	anpInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.updateANP,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+func (c *StatusController) updateCNP(old, cur interface{}) {
+	curCNP := cur.(*secv1alpha1.ClusterNetworkPolicy)
+	oldCNP := old.(*secv1alpha1.ClusterNetworkPolicy)
+	if oldCNP.Status == curCNP.Status {
+		return
+	}
+	key := internalNetworkPolicyKeyFunc(oldCNP)
+	c.queue.Add(key)
+}
+
+func (c *StatusController) updateANP(old, cur interface{}) {
+	curANP := cur.(*secv1alpha1.NetworkPolicy)
+	oldANP := old.(*secv1alpha1.NetworkPolicy)
+	if oldANP.Status == curANP.Status {
+		return
+	}
+	key := internalNetworkPolicyKeyFunc(oldANP)
+	c.queue.Add(key)
+}
+
+func (c *StatusController) UpdateStatus(status *controlplane.NetworkPolicyStatus) error {
+	key := status.Name
+	_, found, _ := c.internalNetworkPolicyStore.Get(key)
+	if !found {
+		klog.Infof("NetworkPolicy %s has been deleted, skip updating its status", key)
+		return nil
+	}
+	func() {
+		c.statusesLock.Lock()
+		defer c.statusesLock.Unlock()
+		statusPerNode, exists := c.statuses[key]
+		if !exists {
+			statusPerNode = map[string]*controlplane.NetworkPolicyNodeStatus{}
+			c.statuses[key] = statusPerNode
+		}
+		for i := range status.Nodes {
```

Comment:
Even one Node updates the status, we need to iterate all Nodes?

Reply:
That's true. However, in the spirit of controller, even many all Nodes update their status multiple times before the policy gets processed, so it doesn't happen that each Node update triggers a full iteration.